addons/utm/models/utm.py

```diff
@@ -42,14 +47,62 @@ def _group_expand_stage_ids(self, stages, domain, order):
         stage_ids = stages._search([], order=order, access_rights_uid=SUPERUSER_ID)
         return stages.browse(stage_ids)
 
+    def _merge_utm_campaigns(self, campaign_to_keep, merge_values={}):
+        """Merge two or more campaigns. Archive all campaigns except campaign_to_keep, and set a reference 
+        to campaign_to_keep for all archived campaigns to lead statistics to the right campaign."""
+        if len(self) <= 1:
+            raise UserError(_('Please select more than one campaign from the list.'))
+        if not all(campaign.active for campaign in self):
+            raise UserError(_('Only active campaigns can be merged.'))
+        if not campaign_to_keep or campaign_to_keep not in self:
+            raise UserError(_('Kept campaign should be one of the merged campaigns.'))
+
+        merge_values['is_website'] = any(campaign.is_website for campaign in self)
+        campaign_to_keep.write(merge_values)
+
+        deactivated_campaigns = self - campaign_to_keep
+        deactivated_campaigns.write({'active': False, 'reference_utm_campaign_id': campaign_to_keep.id})
+        self.search([('reference_utm_campaign_id', 'in', deactivated_campaigns.ids)]).write({
+            'reference_utm_campaign_id': campaign_to_keep.id
+        })
+
+        self._clean_merged_campaigns_references(campaign_to_keep)
+
+    def _clean_merged_campaigns_references(self, merged_campaign):
+        """After a campaign merge, other modules with m2o linked to deactivated campaign should be redirect 
+        to the merged one. Search all fields that is m2o and linked to utm.campaign and not related field. 
+        Also ignore fields belong to AbstractModel and models we intentionally ignored."""
+        fields_to_check = self.env['ir.model.fields'].search([
+            ('ttype', '=', 'many2one'), 
+            ('relation', '=', 'utm.campaign'), 
+            ('related', '=', False)])
+        for field in fields_to_check:
+            if not field.model_id.transient:
+                model = self.env[field.model]
+                if model._auto and model._name not in self._get_ignored_merge_models():
+                    self._clean_merged_campaigns_reference(model._name, field.name, merged_campaign)
+
+    def _clean_merged_campaigns_reference(self, model_name, field_name, merged_campaign):
+        """Given a model_name and field_name of its m2o field link to utm.campaign, change all records link to 
+        archived campaigns to the one we keep."""
+        model = self.env[model_name]
+        record_to_redirect = model.search([(field_name, 'in', (self - merged_campaign).ids)])
```

Comment:
The user searching on the model might not have the necessary access rights.

I think this should be sudoed as well.